homeassistant/components/sensor/solaredge.py

```diff
@@ -0,0 +1,161 @@
+}
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Required(CONF_SITE_ID): cv.string,
+    vol.Optional(CONF_MONITORED_VARIABLES, default=[]):
+    vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)])
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_entities,
+                         discovery_info=None):
+    """Create the SolarEdge Monitoring API sensor."""
+    api_key = config.get(CONF_API_KEY, None)
+    site_id = config.get(CONF_SITE_ID, None)
+
+    # Create solaredge data service which will retrieve and update the data.
+    data = SolarEdgeData(hass, api_key, site_id)
+
+    # Create a new sensor for each sensor type.
+    entities = []
+    for sensor_type in config[CONF_MONITORED_VARIABLES]:
+        sensor = SolarEdgeSensor(sensor_type, data)
+        entities.append(sensor)
+
+    async_add_entities(entities, True)
+
+    # Schedule first data service update straight away.
+    async_track_point_in_utc_time(hass, data.async_update, dt_util.utcnow())
+
+
+class SolarEdgeSensor(Entity):
+    """Representation of an SolarEdge Monitoring API sensor."""
+
+    def __init__(self, sensorType, data):
+        """Initialize the sensor."""
+        self.type = sensorType
+        self.data = data
+
+        self._name = SENSOR_TYPES[self.type][0]
+        self._unit_of_measurement = SENSOR_TYPES[self.type][1]
+
+    @property
+    def name(self):
+        """Return the name."""
+        return 'solaredge_' + self.type
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._unit_of_measurement
+
+    @property
+    def icon(self):
+        """Return the sensor icon."""
+        return SENSOR_TYPES[self.type][2]
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        if self.type in self.data.data:
+            return self.data.data.get(self.type)
+        return 0
+
+
+class SolarEdgeData:
+    """Get and update the latest data."""
+
+    def __init__(self, hass, api_key, site_id):
+        """Initialize the data object."""
+        self.hass = hass
+        self.data = {}
+        self.api_key = api_key
+        self.site_id = site_id
+
+    @asyncio.coroutine
+    def schedule_update(self, minutes):
+        """Schedule an update after minute minutes."""
+        nxt = dt_util.utcnow() + timedelta(minutes=minutes)
+        _LOGGER.debug("Scheduling next SolarEdge update in %s minutes",
+                      minutes)
+        async_track_point_in_utc_time(self.hass, self.async_update, nxt)
+
+    @asyncio.coroutine
+    def async_update(self, *_):
+        """Update the data from the SolarEdge Monitoring API."""
+        import solaredge
+        from requests.exceptions import HTTPError, ConnectTimeout
+
+        # Create new SolarEdge object to retrieve data
+        api = solaredge.Solaredge(self.api_key)
```

Comment:
This should be setup in `async_setup_platform()` and passed down. During `async_setup_platform()` check if the credentials are valid. If not, make the setup fail. This way the users will not end-up with a non-working platform.